receiver/receivertest/contract_checker.go

```diff
@@ -0,0 +1,405 @@
+// The result of the decision function becomes the return value of ConsumeLogs/Trace/Metrics.
+// Supplying different decision functions allows to test different scenarios of the contract
+// between the receiver and it next consumer.
+type consumeDecisionFunc func(ids IDSet) error
+
+var errNonPermanent = errors.New("non permanent error")
+var errPermanent = errors.New("permanent error")
+
+// randomNonPermanentErrorConsumeDecision is a decision function that succeeds approximately
+// half of the time and fails with a non-permanent error the rest of the time.
+func randomNonPermanentErrorConsumeDecision(_ IDSet) error {
+	if rand.Float32() < 0.5 {
+		return errNonPermanent
+	}
+	return nil
+}
+
+// randomPermanentErrorConsumeDecision is a decision function that succeeds approximately
+// half of the time and fails with a permanent error the rest of the time.
+func randomPermanentErrorConsumeDecision(_ IDSet) error {
+	if rand.Float32() < 0.5 {
+		return consumererror.NewPermanent(errPermanent)
+	}
+	return nil
+}
+
+// randomErrorsConsumeDecision is a decision function that succeeds approximately
+// a third of the time, fails with a permanent error the third of the time and fails with
+// a non-permanent error the rest of the time.
+func randomErrorsConsumeDecision(_ IDSet) error {
+	r := rand.Float64()
+	third := 1.0 / 3.0
+	if r < third {
+		return consumererror.NewPermanent(errPermanent)
+	}
+	if r < 2*third {
+		return errNonPermanent
+	}
+	return nil
+}
+
+// mockConsumer accepts or drops the data from the receiver based on the decision made by
+// consumeDecisionFunc and remembers the accepted and dropped data sets for later checks.
+// mockConsumer implements all 3 consume functions: ConsumeLogs/ConsumeTraces/ConsumeMetrics
+// and can be used for testing any of the 3 signals.
+type mockConsumer struct {
+	t                    *testing.T
+	consumeDecisionFunc  consumeDecisionFunc
+	mux                  sync.Mutex
+	acceptedIds          IDSet
+	droppedIds           IDSet
+	nonPermanentFailures int
+}
+
+func (m *mockConsumer) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{}
+}
+
+func (m *mockConsumer) ConsumeTraces(_ context.Context, data ptrace.Traces) error {
+	ids, err := IDSetFromTraces(data)
+	require.NoError(m.t, err)
+	return m.consume(ids)
+}
+
+// IDSetFromTraces computes an IDSet from given ptrace.Traces. The IDSet will contain ids of all spans.
+func IDSetFromTraces(data ptrace.Traces) (IDSet, error) {
+	ds := map[UniqueIDAttrDataType]bool{}
+	rss := data.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		ils := rss.At(i).ScopeSpans()
+		for j := 0; j < ils.Len(); j++ {
+			ss := ils.At(j).Spans()
+			for k := 0; k < ss.Len(); k++ {
+				elem := ss.At(k)
+				key, exists := elem.Attributes().Get(UniqueIDAttrName)
+				if !exists {
+					return ds, fmt.Errorf("invalid data element, attribute %q is missing", UniqueIDAttrName)
+				}
+				if key.Type() != pcommon.ValueTypeInt {
```

Comment:
I'm not sure that all the receivers can produce integer attributes. Maybe we should use string representation for `UniqueIDAttrDataType`?

Reply:
Changed to string.